src/app/ClusterObjectUtils.h

```diff
@@ -0,0 +1,312 @@
+    return writer.Put(tag, x);
+}
+
+CHIP_ERROR Encode(TLV::TLVWriter &writer, uint64_t tag, chip::Span<char> &x) {
+    return writer.PutString(tag, x.data(), (uint32_t)x.size());
+}
+
+template <typename X>
+CHIP_ERROR Encode(TLV::TLVWriter &writer, uint64_t tag, chip::Span<X> &x) {
+    TLV::TLVType type;
+
+    ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Array, type));
+    for (auto item : x) {
+        ReturnErrorOnFailure(ClusterObjectUtils::Encode(writer, TLV::AnonymousTag, item));
+    }
+    ReturnErrorOnFailure(writer.EndContainer(type));
+
+    return CHIP_NO_ERROR;
+}
+
+/*
+ * @brief
+ *
+ * This specific variant that encodes cluster objects (like structs, commands, events) to TLV
+ * depends on the presence of an Encode method on the object to present. The signature of that method
+ * is as follows:
+ *
+ * CHIP_ERROR <Object>::Encode(TLVWriter &writer, uint64_t tag);
+ *
+ *
+ */
+template <typename X, typename std::enable_if<std::is_class<X>::value, X>::type* = nullptr>
+CHIP_ERROR Encode(TLV::TLVWriter &writer, uint64_t tag, X &x) {
+    return x.Encode(writer, tag);
+}
+
+//
+// Decode
+//
+template <typename X, typename std::enable_if<std::is_integral<X>::value, int>::type = 0>
+CHIP_ERROR Decode(TLV::TLVReader &reader, X &x) {
+    return reader.Get(x);
+}
+
+CHIP_ERROR Decode(TLV::TLVReader &reader, chip::ByteSpan &x) {
+    VerifyOrReturnError(reader.GetType() == TLV::kTLVType_ByteString, CHIP_ERROR_UNEXPECTED_TLV_ELEMENT);
+    return reader.Get(x);
+}
+
+CHIP_ERROR Decode(TLV::TLVReader &reader, chip::Span<char> &x) {
+    chip::ByteSpan bs;
+  
+    //
+    // Like for byte string, a span is returned that points into the underlying
+    // buffer in the reader itself.
+    //
+    VerifyOrReturnError(reader.GetType() == TLV::kTLVType_UTF8String, CHIP_ERROR_UNEXPECTED_TLV_ELEMENT); 
+    ReturnErrorOnFailure(reader.Get(bs));
+    x = chip::Span<char>((char *)bs.data(), (size_t)bs.size());
+    return CHIP_NO_ERROR;
+}
+
+/*
+ * @brief
+ *
+ * This specific variant that decodes cluster objects (like structs, commands, events) from TLV
+ * depends on the presence of a Decode method on the object to present. The signature of that method
+ * is as follows:
+ *
+ * CHIP_ERROR <Object>::Decode(TLVReader &reader);
+ *
+ */
+template <typename X, typename std::enable_if<std::is_class<X>::value, X>::type* = nullptr>
+CHIP_ERROR Decode(TLV::TLVReader &reader, X &x) {
+    return x.Decode(reader);
+}
+
+/*
+ * @brief
+ *
+ * This class provides an iteratable decoder of list items within TLV payloads
+ * such that no memory has to be provided ahead of time to store the entirety of the decoded
+ * list contents.
+ *
+ * Specifically, this provides a read-only input iterator that overloads the '->' operator to dynamically
+ * decode the list item at a given position in the stored TLVReader and return it to the caller.
+ *
+ */
+template <typename T>
+class IteratableList {
```

Comment:
So that you can always 'get back to the beginning of the list' if you want to, without having to re-decode the entire object again. This is more in alignment with standard iterator patterns.